### PR TITLE
HAWQ-1036. Implement user impersonation in PXF

### PR DESCRIPTION
**Pull request by denalex**

### CODE REVIEW

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/servlet/SecurityServletFilter.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+ * Listener on lifecycle events of our webapp
+ */
+public class SecurityServletFilter implements Filter {
+
+    private static final Log LOG = LogFactory.getLog(SecurityServletFilter.class);
+    private static final String USER_HEADER = "X-GP-USER";
+    private static final String MISSING_HEADER_ERROR = String.format("Header %s is missing in the request", USER_HEADER);
+    private static final String EMPTY_HEADER_ERROR = String.format("Header %s is empty in the request", USER_HEADER);
+
+
```
Comment (shivzone) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/servlet/SecurityServletFilter.java]: extra newline

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/servlet/SecurityServletFilter.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+    }
+
+    /**
+     * If user impersonation is configured, examines the request for the presense of the expected security headers
+     * and create a proxy user to execute further request chain. Responds with an HTTP error if the header is missing
+     * or the chain processing throws an exception.
+     *
+     * @param request
```
Comment (shivzone) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/servlet/SecurityServletFilter.java]: Not entirely sure if javadocs will be happy with params without description

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/servlet/SecurityServletFilter.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hawq.pxf.service.utilities.SecureLogin;
+
+import javax.servlet.*;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import java.io.IOException;
+import java.security.PrivilegedExceptionAction;
```
Comment (shivzone) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/servlet/SecurityServletFilter.java]: Order of imports

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/utilities/ProtocolData.java [hunk 4]
```diff
@@ -356,6 +361,20 @@ public float getStatsSampleRatio() {
         return statsSampleRatio;
     }
 
+    private void parseSecurityProperties() {
+        // obtain identity of the end-user -- mandatory only when impersonation is enabled
+        if (SecureLogin.isUserImpersonationEnabled()) {
+            this.user = getProperty("USER");
+        } else {
+            this.user = getOptionalProperty("USER");
```
Comment (shivzone) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/utilities/ProtocolData.java]: Do we need the user property if impersonation is not enabled ?

Comment (denalex) [reply pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/utilities/ProtocolData.java]: the property is part of the InputData, so we fill it up anyways. Disabling impersonation disables running the request in the SecurityContext of the user, but user info might be important for some other things (like logging).

**Pull request closed without merging**